source/common/upstream/upstream_impl.h

```diff
@@ -453,7 +471,7 @@ class HostImpl : public HostDescriptionImpl,
   // TODO(wbpcode): should we store the EDS health status to health_flags_ to get unified status or
   // flag access? May be we could refactor HealthFlag to contain all these statuses and flags in the
   // future.
-  HealthStatus health_status_{};
+  std::atomic<Host::HealthStatus> health_status_{};
```

Comment:
When do we modify health status cross-thread?

Reply:
When the health status is updated by the EDS API.

Reply:
So, the worker threads use this as a way to receive EDS health status from the main thread? I think atomics aren't super ideal here, given we also have other health fields atomics, and their updates might have corner cases where they update in different orders depending on memory ordering model of the machine. It's generally fine for stats and maybe even health status if you have only a single field, but when you have cross-field dependency with `health_flags_`, it's something I'd avoid. Please see https://abseil.io/docs/cpp/atomic_danger.

I wonder if all this shared health status between main thread and worker threads for a host should be under a reader-writer lock, reflecting that this is mostly read and fine to be accessed in parallel.

Reply:
Not sure the performance of the rwlock in multiple thread. 

 🤔 Another option is to store all the eds status in the `health_flags_`.

>   // TODO(wbpcode): should we store the EDS health status to health_flags_ to get unified status or
>  // flag access? May be we could refactor HealthFlag to contain all these statuses and flags in the
>  // future.

Reply:
The lesser evil probably would be to have a single atomic vs. multiple atomics tracking related state. It's still somewhat dangerous as per that link, but probably no worse than today. Is it plausible t oombine them?

Reply:
>  Is it plausible t oombine them.

I think eds unhelthy and eds degrade have been set to the health_flags_. So I think it should be ok. But I am not sure if it's the most efficient way to store these statuses.

Reply:
Why isn't flattening into a single variable the most efficient? In any case, I think we're kind of just stepping around the bigger problem, we don't have a clear set of documented expectations and invariants when it comes to cross-thread health.

Reply:
Because the `health_flags_` is desinged to store multiple statuses at same time. If we use it to store all eds statuses, we need check the flag one by one for get the original status.

```
if (health_flags_ & UKNOW_FLAG) {
  return unkonw;
} else if (....) {
  // ......
}
```

>  we don't have a clear set of documented expectations and invariants when it comes to cross-thread health.

I can add some more comment to the health status interfaces this time.

---

btw, I am thinking if the new `healthStatus()` support has truely value. we add the `healthStatus()` method in the #24094 to expose more detailed host status (Unknow, healthy, unhealthy, timeout, etc).

However, years go, the `coarseHealth()` still enough for users. And considering the bug #27830, maybe no accutally users want the detailed `healthStatus()`. `coarseHealth()` is enough.

Maybe we should stop and go back a step, to remove all this redundant support and just keep the original `Unhealthy`, `Degraded`, `Healthy`.

Reply:
https://github.com/envoyproxy/envoy/issues/27830 seems something to fix in any case but not the deal maker/breaker. How do you conclude that `coarseHealth()` is enough? The reason I ask is that when I approved https://github.com/envoyproxy/envoy/pull/24094 I was under the impression this was going to be available in EDS config dumps - how do we know if they are being used actively or not?

Reply:
1. Because only `coarseHealth()` is used when dump the eds config.
2. Because the `eds_health_status_` never be updated correctly by the EDS API.

And considering the #27830, the only place where use the `healthStatus()` but has bugs, so, I think `healthStatus()` never be used correctly.

Reply:
@markdroth we originally added this more detailed EDS health status for gRPC parity I think? It seemed to make sense at the time, but according to @wbpcode we're not actually using it now it seems.

Reply:
Note that gRPC honors the full-fidelity EDS health status in `override_host_status`.  If Envoy reverts to its old behavior, then gRPC and Envoy will have different behavior, and I don't think that's what we want.

Reply:
cc @markdroth Hi, thanks for the response and explaination.

I think the coarse health is well [documented](https://github.com/envoyproxy/envoy/blob/main/api/envoy/config/core/v3/health_check.proto) and work well for all current Envoy users. And in the Envoy, the fine-grained status support never be used or never work correctly.

I may want drop the fine-grained status support and re-open #23213 **until actual Envoy users complain about the health status management.**

Although this will result in minor behavior difference between gRPC proxy and Envoy, but I think it's better than introducing new complexity to add feature that senseless for Envoy users.

Reply:
> I think the coarse health is well [documented](https://github.com/envoyproxy/envoy/blob/main/api/envoy/config/core/v3/health_check.proto)

That file defines the fine-grained EDS health status values.  I don't see anything there that actually documents the coarse-grained values or the fact that data planes are expected to convert the fine-grained values into coarse-grained values internally.

But more fundamentally, I don't think this is a documentation issue; I think it's an issue of the defined meaning of the xDS API.  The xDS protos have an enum that defines the fine-grained EDS health status values, and the `override_host_status` field uses those same enum values.  The only reasonable behavior for this API is that we honor the individual values.  Any other behavior is completely counter-intuitive from the perspective of the way these APIs are defined in xDS.

I also don't think it's okay for Envoy and gRPC to have semantically different behavior here.  If any xDS-aware data plane just ignores what the API says whenever it's convenient, then the xDS ecosystem is not going to provide any actual definition of data plane behavior.

Reply:
Am I misunderstanding? I think the document declare this point clearly? In the Envoy, the `UNKNOW` will be interpreted as `HEALTHY` and `DRAINING`/`TIMEOUT` will be interpreted as `UNHEALTHY`?

> // The health status is not known. This is interpreted by Envoy as ``HEALTHY``.              <-- here

>   // Connection draining in progress. E.g.,
>  // `<https://aws.amazon.com/blogs/aws/elb-connection-draining-remove-instances-from-service-with-care/>`_
>  // or
>  // `<https://cloud.google.com/compute/docs/load-balancing/enabling-connection-draining>`_.
>  // This is interpreted by Envoy as ``UNHEALTHY``.                                                           <-- here

>   // Health check timed out. This is part of HDS and is interpreted by Envoy as
>  // ``UNHEALTHY``                                                                                                              <-- here

Reply:
The documentation says what Envoy considers healthy or unhealthy, but it doesn't actually say what that means -- i.e., what specific behavior will be different based on that decision of healthiness?  And I don't think anything there implies that, for example, setting `override_host_status` to `DRAINING` will actually match hosts in `UNHEALTHY` state.  Those two states are not the same thing, and there will be cases where users will want to match one without the other, and I don't see anything in the documentation that would lead a reader to expect the counter-intuitive behavior of matching statuses that are not actually in the list.

In any case, as I said above, the problem here is not just documentation; the problem is that the coarse-grained behavior is inherently broken.  If `override_host_status` was expressed in terms of coarse-grained health status, then you might be able to argue that Envoy's old behavior made sense -- although it would still probably not be fine-grained enough to address the kind of use-case I cited above.  But since it's actually expressed using the fine-grained EDS health status, I think the natural expectation of this API is that it will actually match endpoints in the specified statuses.  I think any other behavior is simply broken with respect to the API semantics.

Reply:
Thanks for your patient explanation. I am not convinced because I think the document is clearly enough to tell users that Envoy has specifc logic at there.

But I admit the #23213 is a reasonable bug. But now, I don't think it has no big value to be fixed before some actual users to complain that.

But then again, I didn't acutally insist on my point in particular. I just want to fix #27830 (which acutally influence our users) before we cut 1.27.

So I think we should make a decision as soon as possible because we may cut the new release these days. If @markdroth strongly object to drop the fine-grained status support, a quick fix (store the eds status by additional atomic int and update in the main thread, I don't think there are any side effects other than the possibility of inconsistent health status in some very extreme cases) is also acceptable to me.